### PR TITLE
Package batsat.0.7

### PR DESCRIPTION
**Pull request by c-cube**
### `batsat.0.7`
OCaml bindings for batsat, a SAT solver in rust

---
* Homepage: https://github.com/c-cube/batsat-ocaml/
* Source repo: git+https://github.com/c-cube/batsat-ocaml.git
* Bug tracker: https://github.com/c-cube/batsat-ocaml/issues

---
:camel: Pull-request generated by opam-publish v2.1.0

### CODE REVIEW

File: packages/batsat/batsat.0.7/opam [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+maintainer: "simon.cruanes.2007@m4x.org"
+build: [
+  ["./build_rust.sh"]
+  ["dune" "build" "-p" name "-j" jobs]
+  ["dune" "build" "@doc" "-p" name "-j" jobs] {with-doc}
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+# TODO: add dependency on cargo?
+depends: [
+  "ocaml" { >= "4.06.0" }
```
Comment (mseri) [inline packages/batsat/batsat.0.7/opam:13]:
```suggestion
  "ocaml" { >= "4.07.0" }
```
I think this is needed for
```
#=== ERROR while compiling batsat.0.7 =========================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.06.1 | pinned(https://github.com/c-cube/batsat-ocaml/archive/v0.7.tar.gz)
# path                 ~/.opam/4.06/.opam-switch/build/batsat.0.7
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p batsat -j 31
# exit-code            1
# env-file             ~/.opam/log/batsat-1695-ffb3fd.env
# output-file          ~/.opam/log/batsat-1695-ffb3fd.out
### output ###
#       ocamlc src/icnf-solve/.icnf_solve.eobjs/byte/dune__exe__Icnf_solve.{cmi,cmo,cmt}
# File "src/icnf-solve/Icnf_solve.ml", line 46, characters 4-14:
# Warning 3: deprecated: module CCOpt
# use CCOption instead
#     ocamlopt src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Icnf_solve.{cmx,o}
# File "src/icnf-solve/Icnf_solve.ml", line 46, characters 4-14:
# Warning 3: deprecated: module CCOpt
# use CCOption instead
#     ocamlopt tests/test1.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlopt.opt -w -40 -w @8 -safe-string -g -o tests/test1.exe /home/opam/.opam/4.06/lib/ocaml/unix.cmxa -I /home/opam/.opam/4.06/lib/ocaml /home/opam/.opam/4.06/lib/ocaml/threads/threads.cmxa -I /home/opam/.opam/4.06/lib/ocaml src/batsat.cmxa -I src tests/.test1.eobjs/native/dune__exe__Test1.cmx)
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:67: multiple definition of `caml_major_ring'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:67: first defined here
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:68: multiple definition of `caml_major_ring_index'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:68: first defined here
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:69: multiple definition of `caml_major_work_credit'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:69: first defined here
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking
#     ocamlopt src/icnf-solve/icnf_solve.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.06/bin/ocamlopt.opt -w -40 -safe-string -warn-error -a+8 -g -o src/icnf-solve/icnf_solve.exe /home/opam/.opam/4.06/lib/seq/seq.cmxa /home/opam/.opam/4.06/lib/either/either.cmxa /home/opam/.opam/4.06/lib/containers/monomorphic/containers_monomorphic.cmxa /home/opam/.opam/4.06/lib/containers/containers.cmxa /home/opam/.opam/4.06/lib/ocaml/unix.cmxa -I /home/opam/.opam/4.06/lib/ocaml /home/opam/.opam/4.06/lib/ocaml/threads/threads.cmxa -I /home/opam/.opam/4.06/lib/ocaml src/batsat.cmxa -I src src/icnf-solve/.icnf_solve.eobjs/native/dune__exe.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Common.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Lexer.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Solver.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Solver_minisat.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Icnf_solve.cmx)
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:67: multiple definition of `caml_major_ring'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:67: first defined here
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:68: multiple definition of `caml_major_ring_index'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:68: first defined here
# /usr/bin/ld: /home/opam/.opam/4.06/lib/ocaml/libasmrun.a(major_gc.o):/home/opam/.opam/4.06/.opam-switch/build/ocaml-base-compiler.4.06.1/asmrun/../byterun/caml/major_gc.h:69: multiple definition of `caml_major_work_credit'; src/liblibbatsat.a(boxroot.o):/home/opam/.opam/4.06/lib/ocaml/caml/major_gc.h:69: first defined here
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking
```

File: packages/batsat/batsat.0.7/opam [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  "dune" {>= "2.6"}
+  "odoc" {with-doc}
+  "containers" {with-test}
+  "conf-rust-2018" {build}
+]
+tags: [ "minisat" "solver" "SAT" ]
```
Comment (mseri) [inline packages/batsat/batsat.0.7/opam:19]:
```suggestion
available: arch != "arm32"
tags: [ "minisat" "solver" "SAT" ]
```
Due to
```
#=== ERROR while compiling batsat.0.7 =========================================#
# context              2.1.2 | linux/arm32 | ocaml-base-compiler.4.13.1 | pinned(https://github.com/c-cube/batsat-ocaml/archive/v0.7.tar.gz)
# path                 ~/.opam/4.13/.opam-switch/build/batsat.0.7
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p batsat -j 159
# exit-code            1
# env-file             ~/.opam/log/batsat-723-d88365.env
# output-file          ~/.opam/log/batsat-723-d88365.out
### output ###
#       ocamlc src/icnf-solve/.icnf_solve.eobjs/byte/dune__exe__Icnf_solve.{cmi,cmo,cmt}
# File "src/icnf-solve/Icnf_solve.ml", line 46, characters 4-14:
# 46 |     CCOpt.iter (Vec.push self.vec) first;
#          ^^^^^^^^^^
# Alert deprecated: module CCOpt
# use CCOption instead
#     ocamlopt src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Icnf_solve.{cmx,o}
# File "src/icnf-solve/Icnf_solve.ml", line 46, characters 4-14:
# 46 |     CCOpt.iter (Vec.push self.vec) first;
#          ^^^^^^^^^^
# Alert deprecated: module CCOpt
# use CCOption instead
#     ocamlopt tests/test1.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -w @8 -safe-string -g -o tests/test1.exe /home/opam/.opam/4.13/lib/ocaml/unix.cmxa -I /home/opam/.opam/4.13/lib/ocaml /home/opam/.opam/4.13/lib/ocaml/threads/threads.cmxa -I /home/opam/.opam/4.13/lib/ocaml src/batsat.cmxa -I src tests/.test1.eobjs/native/dune__exe__Test1.cmx)
# /usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/libgcc.a(_arm_addsubdf3.o): in function `__floatundidf':
# (.text+0x30c): multiple definition of `__aeabi_ul2d'; src/liblibbatsat.a(compiler_builtins-441c17dbb4fcc13c.compiler_builtins.80nri95o-cgu.100.rcgu.o):/usr/src/rustc-1.48.0/vendor/compiler_builtins/src/macros.rs:225: first defined here
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
#     ocamlopt src/icnf-solve/icnf_solve.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -safe-string -warn-error -a+8 -g -o src/icnf-solve/icnf_solve.exe /home/opam/.opam/4.13/lib/either/either.cmxa /home/opam/.opam/4.13/lib/containers/monomorphic/containers_monomorphic.cmxa /home/opam/.opam/4.13/lib/containers/containers.cmxa /home/opam/.opam/4.13/lib/ocaml/unix.cmxa -I /home/opam/.opam/4.13/lib/ocaml /home/opam/.opam/4.13/lib/ocaml/threads/threads.cmxa -I /home/opam/.opam/4.13/lib/ocaml src/batsat.cmxa -I src src/icnf-solve/.icnf_solve.eobjs/native/dune__exe.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Common.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Lexer.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Solver.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Solver_minisat.cmx src/icnf-solve/.icnf_solve.eobjs/native/dune__exe__Icnf_solve.cmx)
# /usr/bin/ld: /usr/lib/gcc/arm-linux-gnueabihf/10/libgcc.a(_arm_addsubdf3.o): in function `__floatundidf':
# (.text+0x30c): multiple definition of `__aeabi_ul2d'; src/liblibbatsat.a(compiler_builtins-441c17dbb4fcc13c.compiler_builtins.80nri95o-cgu.100.rcgu.o):/usr/src/rustc-1.48.0/vendor/compiler_builtins/src/macros.rs:225: first defined here
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
```

File: packages/batsat/batsat.0.7/opam [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  ["dune" "runtest" "-p" name "-j" jobs] {with-test}
+]
+# TODO: add dependency on cargo?
+depends: [
+  "ocaml" { >= "4.06.0" }
+  "dune" {>= "2.6"}
+  "odoc" {with-doc}
+  "containers" {with-test}
+  "conf-rust-2018" {build}
+]
```
Comment (mseri) [inline packages/batsat/batsat.0.7/opam:18]:
```suggestion
]
conflicts: [ "base-domains" ]
```
See: https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/d231e6a97b709b57ba2a39d113577e08c77d6bd3/variant/opam-2.1,extras,domains,batsat.0.7

**Comment by mseri** [comment]
@c-cube debian and s390x are failing due to rust being too old, the arm32 and domains failures are still there. I think the easiest thing is to make the two changes suggested above

File: packages/batsat/batsat.0.7/opam (hunk 2 above)
Comment (mseri) [reply packages/batsat/batsat.0.7/opam:19]: This is solved now

File: packages/batsat/batsat.0.7/opam (hunk 1 above)
Comment (mseri) [reply packages/batsat/batsat.0.7/opam]: This is solved now

**Comment by mseri** [comment]
All failures are due to too old rust versions, this is good, thanks!

**Comment by c-cube** [comment]
thank you!

**Pull request merged**